### PR TITLE
Add System.CodeDom collections tests

### PR DESCRIPTION
**Pull request by hughbe**
Part of #12179

/cc @stephentoub

### CODE REVIEW

**Comment by hughbe** [comment]
Seems like I've fluffed up the formatting again, will fix soon

**Comment by stephentoub** [comment]
> Seems like I've fluffed up the formatting again, will fix soon

Thanks, @hughbe. Please use spaces rather than tabs.

File: src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs [hunk 1]
```diff
@@ -146,9 +152,20 @@ public void ICollection_NonGeneric_SyncRoot_NonNull(int count)
...
         [Theory]
         [MemberData(nameof(ValidCollectionSizes))]
+        public void ICollection_NonGeneric_SyncRoot_Null(int count)
+        {
+            if (ICollection_NonGeneric_SupportsSyncRoot && ICollection_NonGeneric_HasNullSyncRoot)
+            {
+                ICollection collection = NonGenericICollectionFactory(count);
+                Assert.Null(collection.SyncRoot);
+            }
+        }
```
Comment (stephentoub) [inline src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs]: Rather than introducing a new test, I think I'd prefer if they were a single test that verified the SyncRoot was what it was supposed to be, e.g.
```C#
public void ICollection_NonGeneric_SyncRoot_NullOrNonNull(int count)
{
    if (!ICollection_NonGeneric_SupportsSyncRoot) return;

    ICollection collection = NonGenericICollectionFactory(count);
    Assert.Equal(ICollection_NonGeneric_HasNullSyncRoot, collection.SyncRoot == null);
}
```

File: src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs [hunk 2]
```diff
@@ -146,9 +152,20 @@ public void ICollection_NonGeneric_SyncRoot_NonNull(int count)
...
+
+        [Theory]
+        [MemberData(nameof(ValidCollectionSizes))]
         public void ICollection_NonGeneric_SyncRootConsistent(int count)
         {
-            if (ICollection_NonGeneric_SupportsSyncRoot)
+            if (ICollection_NonGeneric_SupportsSyncRoot && !ICollection_NonGeneric_HasNullSyncRoot)
```
Comment (stephentoub) [inline src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs]: Is this change needed?  Looks like the test will work even if the collection returns null from SyncRoot.  If we're going to run the test anyway, we may as well do something with it.

File: src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs [hunk 3]
```diff
@@ -196,7 +213,7 @@ public void ICollection_NonGeneric_SyncRoot_MatchesExpectedType(int count)
         [MemberData(nameof(ValidCollectionSizes))]
         public void ICollection_NonGeneric_SyncRoot_ThrowsNotSupportedException(int count)
         {
-            if (!ICollection_NonGeneric_SupportsSyncRoot)
+            if (!ICollection_NonGeneric_SupportsSyncRoot && !ICollection_NonGeneric_HasNullSyncRoot)
```
Comment (stephentoub) [inline src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs]: Rather than having this be a separate test, I'd similarly rather have it be in the single test that verifies the SyncRoot is what we expect it to be: null, non-null, or not even supported.  I realize you didn't introduce this particular test.  I'm just cognizant of the fact that every new test we add here gets multiplied significantly, as it's a theory that's then pulled into a bunch of different tests for a bunch of different instantiations for a bunch of different collection types.  If a test really is verifying something unique, then it's fine to stand alone.  Here, though, this and the other two earlier tests are really all checking the same thing: does SyncRoot return what we expect (if we expect anything).

File: src/Common/tests/System/Collections/IList.NonGeneric.Tests.cs [hunk 4]
```diff
@@ -863,19 +865,34 @@ public void IList_NonGeneric_Remove_NullNotContainedInCollection(int count)
...
-        }
-
-        [Theory]
+		}
+
+		[Theory]
+		[MemberData(nameof(ValidCollectionSizes))]
+		public void IList_NonGeneric_Remove_NonNullNotContainedInCollection_Throws(int count)
+		{
+			if (IList_NonGeneric_RemoveNonExistent_Throws)
```
Comment (stephentoub) [inline src/Common/tests/System/Collections/IList.NonGeneric.Tests.cs]: Same comment as earlier.  I'd prefer we not introduce a new test for this and just add a conditional to the previous one, e.g.
```C#
if (IList_NonGeneric_RemoveNonExistent_Throws)
{
    Assert.Throws<ArgumentException>(() => list.Remove(value));
}
else
{
    list.Remove(value);
    Assert.Equal(count, list.Count);
}
```
We otherwise end up significantly increasing the number of test cases, but half of them end up being nops for a given collection.

File: src/System.CodeDom/src/System/CodeDom/Compiler/TempFiles.cs [hunk 5]
```diff
@@ -112,7 +112,6 @@ private void EnsureTempNameCreated()
                     _basePath = Path.Combine(
                         string.IsNullOrEmpty(TempDir) ? Path.GetTempPath() : TempDir,
                         Path.GetFileNameWithoutExtension(Path.GetRandomFileName()));
-                    string full = Path.GetFullPath(_basePath);
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/TempFiles.cs]: Good spot on the dead code.  This had been part of a permission demand, which was removed, and I neglected to remove the code that was providing the input to the demand.

File: src/System.CodeDom/src/System/CodeDom/Compiler/TempFiles.cs [hunk 6]
```diff
@@ -123,6 +122,10 @@ private void EnsureTempNameCreated()
                     catch (IOException)
                     {
                         retryCount--;
+                        if (retryCount == 0)
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/TempFiles.cs]: Given how you found this, I think we actually need to put back the second half of the condition as well.  If a non-existent directory is provided, we don't want to try 5000 different files, all of which will fail, before actually failing: we want to fail immediately.  But we also can't put it back exactly as it was, as it won't work cross-platform, due to different .  I suggest this:
```C#
if (retryCount == 0 || e is DirectoryNotFoundException)
{
    throw;
}
```
It's not 100% the same, but it addresses the most likely case (unfortunately there's no strongly-typed "file already exists" exception we can special case).  If you can think of other exceptions that we'd get here, we can add those, too.

Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/TempFiles.cs]: (Thanks for finding the issue, btw.)

File: src/System.CodeDom/tests/CodeCollections/CodeAttributeArgumentCollectionTests.cs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		public void Ctor_IsEmpty()
+		{
+			var collection = new CodeAttributeArgumentCollection();
+			Assert.Equal(0, collection.Count);
+		}
+
+		public static IEnumerable<object[]> AddRange_TestData()
+		{
+			yield return new object[] { new CodeAttributeArgument[0] };
+			yield return new object[] { new CodeAttributeArgument[] { new CodeAttributeArgument() } };
```
Comment (stephentoub) [inline src/System.CodeDom/tests/CodeCollections/CodeAttributeArgumentCollectionTests.cs]: Since this is about adding a range, let's have at least three cases: 0, 1, and more than 1.  (If you've added similar tests later in the PR, same feedback applies).

**Comment by stephentoub** [comment]
> This is because each collection uses CollectionBase, so defines its own methods that are strongly typed. As such, there is no way to have common tests that run generically for all of the collections. The IList and ICollection helpers cannot help here, as none of the collections actually implement IList/ICollection

How about defining your own test base class that provides virtual methods for Add, Remove, Ctor, CtorWithArray, CtorWithCollection, etc., and then write the tests in terms of those and have each collection override those methods?  Then most of the collections should be able to share the same set of tests.

**Comment by stephentoub** [comment]
Also, could you separate out the TempFileCollection fix and associated tests into its own PR?  That'll make it easier to squash the general collection test additions while keeping the fix separate.

**Comment by stephentoub** [comment]
Thanks for adding these, @hughbe.  I reviewed the TempFileCollection change and tests, as well as one of the other collection's tests... I'll hold off on reviewing more until you decide what to do about all of the duplication.

File: src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs (hunk 1 above)
Comment (hughbe) [inline src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs]: Addressed in #12680

File: src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs (hunk 3 above)
Comment (hughbe) [inline src/Common/tests/System/Collections/ICollection.NonGeneric.Tests.cs]: Addressed in #12680

File: src/Common/tests/System/Collections/IList.NonGeneric.Tests.cs (hunk 4 above)
Comment (hughbe) [inline src/Common/tests/System/Collections/IList.NonGeneric.Tests.cs]: Addressed in #12680

File: src/System.CodeDom/tests/CodeCollections/TempFileCollectionTests.cs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+				Assert.Equal(fileExists && keepFile, File.Exists(filePath));
+			}
+			finally
+			{
+				if (File.Exists(filePath))
+				{
+					File.Delete(filePath);
+				}
+			}
+        }
```
Comment (hughbe) [inline src/System.CodeDom/tests/CodeCollections/TempFileCollectionTests.cs]: Addressed in #12680

**Comment by hughbe** [comment]
This PR now relies on #12680. I'll work on trying to avoid duplication in private then push when that PR is merged :)
Thanks for your comments :)

**Comment by hughbe** [comment]
You were right - I managed to get rid of bunch of duplicated code. thanks for the suggestion!

File: src/System.CodeDom/tests/CodeCollections/CodeCollectionTestBase.cs [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.CodeDom.Tests
+{
+    public abstract class CodeCollectionTestBase<TCollection, TItem> where TCollection: class where TItem : class, new()
+    {
+        public abstract TCollection Ctor();
```
Comment (stephentoub) [inline src/System.CodeDom/tests/CodeCollections/CodeCollectionTestBase.cs]: Nit: Do all of the collections have a parameterless ctor?  If so, we could use a new() constraint on TCollection as well to avoid needing this abstract method.

**Comment by stephentoub** [comment]
Looks good, @hughbe, and glad to see much of the duplication go away.  Thanks for adding these.
